src/canonical_combining_class.rs

```diff
@@ -0,0 +1,67 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use ucd_parse::{self, UnicodeData};
+
+use crate::args::ArgMatches;
+use crate::error::Result;
+use crate::util::{print_property_values, PropertyValues};
+
+pub fn command(args: ArgMatches<'_>) -> Result<()> {
+    let dir = args.ucd_dir()?;
+    let propvals = PropertyValues::from_ucd_dir(&dir)?;
+    let rows: Vec<UnicodeData> = ucd_parse::parse(&dir)?;
+    let ccc_name = |ccc: u8| {
+        propvals.canonical("canonicalcombiningclass", &ccc.to_string())
+    };
+
+    // If we were tasked with listing the available categories, then do that
+    // and quit.
+    if args.is_present("list-classes") {
+        return print_property_values(&propvals, "Canonical_Combining_Class");
+    }
+
+    // Collect each canonical combining class into an ordered set.
+    let mut name_map: BTreeMap<isize, String> = BTreeMap::new();
+    let mut by_name: BTreeMap<String, BTreeSet<u32>> = BTreeMap::new();
+    let mut assigned = BTreeSet::new();
+    for row in rows {
+        assigned.insert(row.codepoint.value());
+        let ccc_value = row.canonical_combining_class;
+        let ccc_name = ccc_name(ccc_value)?;
+        name_map.entry(ccc_value as isize).or_insert(ccc_name.clone());
```

Comment:
Perhaps use `or_insert_with` to avoid cloning if not needed:

```suggestion
        name_map.entry(ccc_value as isize).or_insert_with(|| ccc_name.clone());
```

Reply:
Oh, good point, thanks!